### ISSUE TITLE
Formatting, for example font color, is lost when all content is removed / Preserve font styles / Change default font

### ISSUE THREAD

**Issue by Steveiwonder** [ckeditor/ckeditor5#10517]
## 📝 Provide detailed reproduction steps (if any)

1. Enter some text into the content area
2. Apply some formatting, change font size, color etc.
3. Delete all that content
4. Type new content

### ✔️ Expected result

I expect to see the newly typed content with the last selected font styles (similar to word, powerpoint, etc)

### ❌ Actual result

The font has the editors default styles applied

### ❓ Possible solution

Don't remove the last span element that is the document editor, as this has the style

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

**Comment by Steveiwonder** [comment]
2 people want this fixed, must be important.

**Comment by Reinmar** [comment]
Why do you think it should work this way?

A common use case for removing all content is to start from scratch. Why keeping the formatting then?

**Comment by Steveiwonder** [comment]
Hey, if you have formatted text in the editor, select all or some, then start typing, you'd expect styling to be maintained, this is pretty common. Try it out in Word, PowerPoint, TinyMCE, Quill, DevExpress, Summernote, Google Docs and many others.

I have a solution which seems to behave in all of our use cases for anyone that would like to use it.

``` javascript
import Plugin from "@ckeditor/ckeditor5-core/src/plugin";
import MoveOperation from "@ckeditor/ckeditor5-engine/src/model/operation/moveoperation";
import Text from "@ckeditor/ckeditor5-engine/src/model/text";
export default class PreserveFontStyle extends Plugin {
  init() {
    this.editor.model.document.on("change:data", (evt, data) => {
      const operations = data.operations?.filter(
        (x) => x instanceof MoveOperation
      );
      if (!operations || operations.length <= 0) {
        return;
      }
      try {
        // grab last move operation
        const moveOperation = operations[operations.length - 1];

        // ensure we are moving from "main" to the "graveyard"
        if (
          moveOperation.sourcePosition.root.rootName != "main" ||
          moveOperation.targetPosition.root.rootName != "$graveyard"
        ) {
          return;
        }

        // get the first node and ensure it's a text node
        const deletedTextNode = moveOperation.targetPosition.nodeAfter;
        if (!(deletedTextNode instanceof Text)) {
          return;
        }
        // get all attributes to re-apply
        const attrs = [...deletedTextNode.getAttributes()];
        if (attrs.length == 0) {
          return;
        }
        // now we need to check for the paragraph is empty, if it is, we insert an invisible character
        // inserting this invisible character maintains styles
        const parentParagraph = moveOperation.sourcePosition?.parent;
        if (
          parentParagraph?.name == "paragraph" &&
          parentParagraph.childCount == 0
        ) {
          this.editor.model.change((writer) => {
            writer.insertText("\u200B", attrs, parentParagraph, 0);
          });
          return;
        }

        const insertedTextNode = moveOperation.sourcePosition.nodeAfter;
        if (!insertedTextNode) {
          return;
        }

        // when the parent paragraph has more than one child, do nothing
        // prevents merging
        if (insertedTextNode.parent.childCount != 1) {
          return;
        }
        this.editor.model.change((writer) => {
          writer.setAttributes(attrs, insertedTextNode);
        });
      } catch (e) {
        console.debug(e);
      }
    });
  }
}
```

**Comment by Steveiwonder** [comment]
Another issue with the same problem: https://github.com/ckeditor/ckeditor5/issues/9946

**Comment by dkrahn** [comment]
Hi @Reinmar, I think it could remove all format if you have no text and then remove text again(with backspace, delete...).
The main reason I see to keep styles is to be able to, for example, change the text(not format) of the whole paragraph. Today you have to keep a character and start typing and only after you can remove the old character.
This may also solve a problem where you change the text alignment the whole text(2 paragraphs) being the last one empty. You will see that the first one aligns correcly but the second one doesn't.

**Comment by CKEditorBot** [comment]
There has been no activity on this issue for the past year. We've marked it as stale and will close it in 30 days. We understand it may be relevant, so if you're interested in the solution, leave a comment or reaction under this issue.

**Comment by hendeltom** [comment]
@dkrahn @Reinmar  That's exactly the situation I ran into... And there's one more point: If you have for example content like this:
```html
<p>
   First line
</p>
<p>
</p>
<p>
   Last line
</p>
```
then you can select all the text and apply font attributes (font family, size etc). But the empty paragraph in the middle is not touched. So if you place the caret there and start typing you'd not have the text formatted with the desired font attributes.
This hard to to explain to our users. For them, it's a bug.

**Comment by arkflpc** [comment]
Research ticket: #19664.

**Comment by arkflpc** [comment]
Let's extend this with the case of removing text in single paragraph.

**Comment by Reinmar** [comment]
Short recap – the "preserve style" thing is a combination of many smaller behaviors that need to be changed.

---

@hendeltom

> then you can select all the text and apply font attributes (font family, size etc). But the empty paragraph in the middle is not touched

This is fixed already :)

---

@dkrahn 

> Hi [@Reinmar](https://github.com/Reinmar), I think it could remove all format if you have no text and then remove text again(with backspace, delete...).
> The main reason I see to keep styles is to be able to, for example, change the text(not format) of the whole paragraph. Today you have to keep a character and start typing and only after you can remove the old character.

@arkflpc, this is the case we IMO focus on here – what happens after you press Backspace that results in emptying a block, regardless of whether the Backspace is pressed on an empty selection or selection of that block (or multiple blocks).

---

Note, the Ctrl+A and Backspace might actually turn out to be a whole different scenario. In fact, I'd prefer if we do not change its behavior. Currently, it removes all the styling and leaves an empty paragraph, even if the first block was something else (e.g. a heading). The reason why I'd keep it working this way is that it allows a full cleanup of the doc to start from scratch. I know that some RTEs insist on keeping the styling but from my experience user preferences aren't obvious here.

---

In other words, let's keep the scope of the changes minimal and address primarily the scenario where you might have  a long section of content styled with e.g. red font, and you suddenly lose that formatting when you alter a content of a paragraph within that section.

**Comment by Witoso** [comment]
> The reason why I'd keep it working this way is that it allows a full cleanup of the doc to start from scratch. I know that some RTEs insist on keeping the styling but from my experience user preferences aren't obvious here.

I will note down one scenario for posterity that invalidates this statement, but it's related to how our editor works in general. The case is about defaults, and in some features we have a cryptic `Default` which may be confusing for users:

<img width="946" height="509" alt="Image" src="https://github.com/user-attachments/assets/fc6be806-c5ba-419e-8833-d49020585516" />

If I ctrl+a and set Arial, the default font in my head is Arial. I think GDoc and Word do have some heuristic for that (afair, if I open a new GDoc, it remembers my preferences from the previous document). Potentially in the future we could have dynamic default changes (font family, font size, etc.) in which these removals feel slightly more logical.
